Make supervisor actor more useful

thomaseizinger opened the pull request:
1. Expose metrics
2. Catch panics

thomaseizinger committed:
Have supervisor track metrics about spawns

thomaseizinger committed:
Restart supervised actors in case they panic

thomaseizinger commented on daemon/src/supervisor.rs:
```diff
@@ -142,6 +192,30 @@ mod tests {
+    async fn supervisor_tracks_panic_metrics() {
+        let _guard = tracing_subscriber::fmt().with_test_writer().set_default();
+
+        std::panic::set_hook(Box::new(|_| ())); // Override hook to avoid panic printing to log.
```
I need to test whether or not we also need to do this in production.

rishflab commented on daemon/src/supervisor.rs:
```diff
@@ -98,3 +111,60 @@ where
+        let (supervisor, task) = supervisor.create(None).run();
+        tokio::spawn(task);
+
+        let metrics = supervisor.send(GetMetrics).await.unwrap();
```
Why does sending a message spawn a new actor?

thomaseizinger replied:
It doesn't! What spawns the actor initially is the `started` lifecycle hook of the supervisor:

https://github.com/itchysats/itchysats/blob/b412ac01c5e5f8387210f39fb02e995414a13626/daemon/src/supervisor.rs#L61-L63

As a consequence of that, the `num_spawns` counter is 1 as soon as the supervisor is started.

rishflab commented on daemon/src/supervisor.rs:
```diff
@@ -13,6 +13,13 @@ pub struct Actor<T, R> {
+
+#[derive(Default, Clone, Copy)]
+pub struct Metrics {
+    /// How many times the supervisor spawned an instance of the actor.
```
looking at the test it seems like like this is the number of messages the actor has handled?

thomaseizinger replied:
Nope, if you look at the implementation, it is the number of times `spawn_new` has been called! :)

rishflab reviewed (commented):
I can't really review the second commit as I have never recorded panics. Left some questions on the first commit.

luckysori commented on daemon/src/supervisor.rs:
```diff
@@ -104,6 +120,29 @@ where
+        }
+
+        self.metrics.num_panics += 1;
+        self.spawn_new(ctx)
```
❓ Why is panicking treated as an automatic restart (as opposed to stopping, which depends on the restart policy)?

thomaseizinger replied:
So I figured that because the cause of the panic is unknown at coding time, it doesn't make much sense to implement a policy.

If I would know the problem already to code the policy, I could also just fix the problem in the first place 😁

The idea of this experiment is that hopefully the panic is temporary and restarting the actor puts it in a known state where we can progress again. Perhaps only for a while if the panic was triggered by recurring data.

So we certainly need to monitor these metrics!

I am thinking of rewriting the trace statements here and putting in a dedicated subscriber. Probably in a follow-up though!

luckysori reviewed (approved):
Cool!

da-kami commented on daemon/src/supervisor.rs:
```diff
@@ -104,6 +120,29 @@ where
+
+        match msg.error.downcast::<String>() {
+            Ok(reason) => {
+                tracing::error!("{actor} panicked: {reason}");
```
Wouldn't it make sense to add the metrics to the logs so we see them?

thomaseizinger replied:
I am thinking about that yeah. Will push an update later!

thomaseizinger replied:
I opted for adding more structured logs, that should make it easier to compute such metrics downstream. Should I remove the metrics again? Only added them for testing really ..

thomaseizinger replied:
I opted for making these things module private for now.

thomaseizinger committed:
Make logs more tracing friendly

By providing these values as fields, we can more easily track them in
log analysis tools.

thomaseizinger committed:
Use `Tasks::add_fallible` API for price feed

This results in us catching one more error: the failed parsing of the
quote. Previously, we just terminated the task without also terminating
the actor. Now we catch all these errors and terminate the actor
afterwards. Hopefully this will make the price feed more resilient.

May fix #1137.
